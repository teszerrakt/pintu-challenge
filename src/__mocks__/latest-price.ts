import { IGetLatestPricePayload } from 'src/apis/latestPrice/interface'

export const MOCK_LATEST_PRICE_DATA: IGetLatestPricePayload[] = [
  {
    pair: 'btc/idr',
    priceChanges: {
      latestPrice: '458103826',
      day: '0.66',
      week: '2.26',
      month: '10.47',
      year: '51.33',
    },
    currency: {
      currencySymbol: 'BTC',
      currencyGroup: 'BTC',
      color: '#F78B1A',
      name: 'Bitcoin',
      logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_BTC.svg',
      listingDate: '2020-09-15T09:43:45Z',
    },
  },
  {
    pair: 'eth/idr',
    priceChanges: {
      latestPrice: '27806628',
      day: '-0.17',
      week: '-1.04',
      month: '-1.83',
      year: '62.84',
    },
    currency: {
      currencySymbol: 'ETH',
      currencyGroup: 'ETH',
      color: '#9011FE',
      name: 'Ethereum',
      logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_ETH.svg',
      listingDate: '2020-09-15T09:43:46Z',
    },
  },
  {
    pair: 'usdt/idr',
    priceChanges: {
      latestPrice: '15002',
      day: '0.04',
      week: '0.33',
      month: '0.14',
      year: '1.11',
    },
    currency: {
      currencySymbol: 'USDT',
      currencyGroup: 'USDT',
      color: '#15A57C',
      name: 'Tether',
      logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_USDT.svg',
      listingDate: '2020-09-15T09:43:47Z',
    },
  },
]
