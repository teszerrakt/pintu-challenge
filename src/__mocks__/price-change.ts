import { IGetPriceChangeResponse } from 'src/interface'

export const MOCK_PRICE_CHANGE_DATA: IGetPriceChangeResponse[] = [
  {
    pair: 'busd/idr',
    latestPrice: '15015',
    day: '0.15',
    week: '0.16',
    month: '0.41',
    year: '1.22',
  },
  {
    pair: '1inch/idr',
    latestPrice: '4876',
    day: '-3.86',
    week: '17.98',
    month: '-16.72',
    year: '-54.89',
  },
  {
    pair: 'alpha/idr',
    latestPrice: '1762',
    day: '-4.13',
    week: '3.34',
    month: '11.66',
    year: '-10.19',
  },
  {
    pair: 'ldo/idr',
    latestPrice: '28919',
    day: '-1.28',
    week: '0.00',
    month: '-2.87',
    year: '203.83',
  },
  {
    pair: 'vet/idr',
    latestPrice: '280',
    day: '2.19',
    week: '26.13',
    month: '-2.10',
    year: '-27.84',
  },
]
