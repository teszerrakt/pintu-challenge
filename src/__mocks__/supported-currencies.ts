import { IGetSupportedCurrenciesResponse } from 'src/interface'

export const MOCK_SUPPORTED_CURRENCIES: IGetSupportedCurrenciesResponse[] = [
  {
    currencyGroup: 'BUSD',
    color: '#FEBF11',
    currencySymbol: 'BUSD',
    name: 'Binance USD',
    logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_BUSD.svg',
    decimal_point: 18,
    listingDate: '2023-02-06T07:00:00Z',
    wallets: [
      {
        currencyGroup: 'BUSD',
        tokenSymbol: 'BUSD',
        decimal_point: 18,
        tokenType: 'ERC-20',
        blockchain: 'Ethereum',
        explorer: 'https://etherscan.io/tx/',
        listingDate: '2023-02-06T07:00:00Z',
        blockchainName: 'Ethereum',
        logo: 'https://s3.ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/blockchain/ERC-20.svg',
      },
      {
        currencyGroup: 'BUSD',
        tokenSymbol: 'BUSDBSC',
        decimal_point: 18,
        tokenType: 'BEP-20',
        blockchain: 'Binance Smartchain',
        explorer: 'https://bscscan.com/tx/',
        listingDate: '2023-02-06T07:00:01Z',
        blockchainName: 'BNB Smart Chain',
        logo: 'https://s3.ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/blockchain/BEP-20+(BSC).svg',
      },
    ],
  },
  {
    currencyGroup: '1INCH',
    color: '#3B76BA',
    currencySymbol: '1INCH',
    name: '1inch',
    logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_1INCH.svg',
    decimal_point: 18,
    listingDate: '2021-08-26T05:15:18Z',
    wallets: [
      {
        currencyGroup: '1INCH',
        tokenSymbol: '1INCH',
        decimal_point: 18,
        tokenType: 'ERC-20',
        blockchain: 'Ethereum',
        explorer: 'https://etherscan.io/tx/',
        listingDate: '2021-08-26T05:15:18Z',
        blockchainName: 'Ethereum',
        logo: 'https://s3.ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/blockchain/ERC-20.svg',
      },
    ],
  },
  {
    currencyGroup: 'ALPHA',
    color: '#1F66E6',
    currencySymbol: 'ALPHA',
    name: 'Alpha Finance',
    logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_ALPHA.svg',
    decimal_point: 18,
    listingDate: '2022-01-13T18:15:20Z',
    wallets: [
      {
        currencyGroup: 'ALPHA',
        tokenSymbol: 'ALPHA',
        decimal_point: 18,
        tokenType: 'ERC-20',
        blockchain: 'Ethereum',
        explorer: 'https://etherscan.io/tx/',
        listingDate: '2022-01-13T18:15:20Z',
        blockchainName: 'Ethereum',
        logo: 'https://s3.ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/blockchain/ERC-20.svg',
      },
    ],
  },
  {
    currencyGroup: 'LDO',
    color: '#469BF0',
    currencySymbol: 'LDO',
    name: 'Lido DAO',
    logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_LDO.svg',
    decimal_point: 18,
    listingDate: '2022-05-20T04:00:18Z',
    wallets: [
      {
        currencyGroup: 'LDO',
        tokenSymbol: 'LDO',
        decimal_point: 18,
        tokenType: 'ERC-20',
        blockchain: 'Ethereum',
        explorer: 'https://etherscan.io/tx/',
        listingDate: '2022-05-20T04:00:18Z',
        blockchainName: 'Ethereum',
        logo: 'https://s3.ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/blockchain/ERC-20.svg',
      },
    ],
  },
  {
    currencyGroup: 'VET',
    color: '#23B2F9',
    currencySymbol: 'VET',
    name: 'VeChain',
    logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_VET.svg',
    decimal_point: 18,
    listingDate: '2022-07-28T07:00:09Z',
    wallets: [
      {
        currencyGroup: 'VET',
        tokenSymbol: 'VET',
        decimal_point: 18,
        tokenType: 'VeChain',
        blockchain: 'VeChain',
        explorer: 'https://explore.vechain.org/transactions/',
        listingDate: '2022-07-28T07:00:09Z',
        blockchainName: 'VeChain',
        logo: 'https://s3.ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/blockchain/Vechain.svg',
      },
    ],
  },
]
