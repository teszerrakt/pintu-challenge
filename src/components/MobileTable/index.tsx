import { ITableProps } from 'src/components/Table'
import { IGetLatestPricePayload } from 'src/apis/latestPrice/interface'
import CryptoLogo from 'src/components/Icons/Logo'
import { stringToCurrency } from 'src/utils/formatter'
import Percentage from 'src/components/Percentage'
import { useState } from 'react'
import Dropdown from 'src/components/Dropdown'

type TFilter = 'day' | 'week' | 'month' | 'year'

function Item({
  data,
  filter,
}: {
  data: IGetLatestPricePayload
  filter: TFilter
}) {
  return (
    <div className="flex items-center p-4 border-b">
      <CryptoLogo url={data.currency.logo} color={data.currency.color} />
      <div className="flex flex-row flex-1 pl-6">
        <div className="flex-1">
          <p className="font-semibold">{data.currency.name}</p>
          <p className="text-sm text-slate-400">
            {data.currency.currencySymbol}
          </p>
        </div>
        <div className="self-end flex-1">
          <p className="font-semibold text-right">
            {stringToCurrency(data.priceChanges.latestPrice)}
          </p>
          <div className="text-sm">
            <Percentage
              value={data.priceChanges[filter]}
              className="justify-end"
            />
          </div>
        </div>
      </div>
    </div>
  )
}

const DROPDOWN_OPTIONS = [
  { key: 'day', value: '24 JAM' },
  { key: 'week', value: '1 MGG' },
  { key: 'month', value: '1 BLN' },
  { key: 'year', value: '1 THN' },
]

function MobileTable({ data }: ITableProps) {
  const [filter, setFilter] = useState<TFilter>('day')

  const handleFilterChange = (value: string) => {
    setFilter(value as TFilter)
  }

  return (
    <div className="flex flex-col">
      <div className="flex items-center justify-between w-full p-4 border-y border-slate-300">
        <p className="text-xs font-semibold ">CRYPTO</p>
        <Dropdown
          options={DROPDOWN_OPTIONS}
          initialValue={filter}
          onChange={handleFilterChange}
        />
      </div>
      <div className="flex flex-col">
        {data.map((item) => (
          <Item key={item.pair} data={item} filter={filter} />
        ))}
      </div>
    </div>
  )
}

export default MobileTable
