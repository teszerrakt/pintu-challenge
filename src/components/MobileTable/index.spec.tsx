import { render, screen, fireEvent } from '@testing-library/react'
import { MOCK_LATEST_PRICE_DATA } from 'src/__mocks__/latest-price'
import MobileTable from 'src/components/MobileTable'

jest.mock('src/components/Icons/Logo')

const mockData = MOCK_LATEST_PRICE_DATA

describe('MobileTable', () => {
  it('renders the table rows correctly', () => {
    render(<MobileTable data={mockData} />)

    const rowElements = screen.getAllByRole('link')
    expect(rowElements.length).toBe(mockData.length)

    const firstRow = rowElements[0]
    expect(firstRow).toMatchSnapshot()
  })

  it('updates the table rows when the filter is changed', () => {
    render(<MobileTable data={mockData} />)

    const selectedFilterOption = screen.getByDisplayValue('24 JAM')

    const initialRowElements = screen.getAllByRole('link')
    expect(initialRowElements.length).toBe(mockData.length)

    fireEvent.change(selectedFilterOption, { target: { value: 'week' } })

    expect(selectedFilterOption).toHaveTextContent('1 MGG')

    const updatedRowElements = screen.getAllByRole('link')
    expect(updatedRowElements.length).toBe(mockData.length)

    const firstRow = updatedRowElements[0]
    expect(firstRow).toHaveTextContent('Bitcoin')
    expect(firstRow).toHaveTextContent('BTC')
  })
})
