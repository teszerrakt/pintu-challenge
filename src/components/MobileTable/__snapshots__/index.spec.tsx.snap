// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileTable renders the table rows correctly 1`] = `
<a
  class="flex items-center p-4 border-b"
  href="undefined/market/BTC"
>
  <div
    class="flex flex-row flex-1 pl-6"
  >
    <div
      class="flex-1"
    >
      <p
        class="font-semibold"
      >
        Bitcoin
      </p>
      <p
        class="text-sm text-slate-400"
      >
        BTC
      </p>
    </div>
    <div
      class="self-end flex-1"
    >
      <p
        class="font-semibold text-right "
      >
        Rp 458.103.826
      </p>
      <div
        class="text-sm"
      >
        <div
          class="flex gap-[5px] text-green-600 justify-end"
          data-testid="percentage"
        >
          <div
            class="movement-icon"
          >
            ▲
          </div>
          <p
            class="font-semibold"
          >
            0.66
            %
          </p>
        </div>
      </div>
    </div>
  </div>
</a>
`;
